openvasd/openvasd.h

```diff
@@ -0,0 +1,312 @@
+
+int
+openvasd_parsed_results (openvasd_connector_t *, unsigned long, unsigned long,
+                         GSList **);
+
+openvasd_resp_t
+openvasd_get_scan_status (openvasd_connector_t *);
+
+openvasd_scan_status_t
+openvasd_parsed_scan_status (openvasd_connector_t *);
+
+int
+openvasd_get_scan_progress (openvasd_connector_t *);
+
+openvasd_resp_t
+openvasd_get_health_alive (openvasd_connector_t *);
+
+openvasd_resp_t
+openvasd_get_health_ready (openvasd_connector_t *);
+
+openvasd_resp_t
+openvasd_get_health_started (openvasd_connector_t *);
+
+/* Scanner preferences */
+
+typedef struct openvasd_param openvasd_param_t;
+
+openvasd_resp_t
+openvasd_get_scan_preferences (openvasd_connector_t *);
+
+int
+openvasd_parsed_scans_preferences (openvasd_connector_t *, GSList **);
+
+void
+openvasd_param_free (openvasd_param_t *);
+
+char *
+openvasd_param_id (openvasd_param_t *);
+
+char *
+openvasd_param_name (openvasd_param_t *);
+
+char *
+openvasd_param_desc (openvasd_param_t *);
+
+int
+openvasd_param_mandatory (openvasd_param_t *);
+
+char *
+openvasd_param_type (openvasd_param_t *);
+
+char *
+openvasd_param_default (openvasd_param_t *);
+
+/* Target builder */
+
+typedef struct openvasd_target openvasd_target_t;
+
+typedef struct openvasd_vt_single openvasd_vt_single_t;
+
+typedef struct openvasd_credential openvasd_credential_t;
+
+openvasd_target_t *
+openvasd_target_new (const gchar *, const gchar *, const gchar *, const gchar *,
+                     int, int);
+
+void
+openvasd_target_set_finished_hosts (openvasd_target_t *, const gchar *);
+
+void
+openvasd_target_add_alive_test_methods (openvasd_target_t *, gboolean, gboolean,
+                                        gboolean, gboolean, gboolean);
+
+void
+openvasd_target_free (openvasd_target_t *);
+
+openvasd_credential_t *
+openvasd_credential_new (const gchar *, const gchar *, const gchar *);
+
+void
+openvasd_credential_set_auth_data (openvasd_credential_t *, const gchar *,
+                                   const gchar *);
+void
+openvasd_credential_free (openvasd_credential_t *);
+
+void
+openvasd_target_add_credential (openvasd_target_t *, openvasd_credential_t *);
+
+openvasd_vt_single_t *
+openvasd_vt_single_new (const gchar *);
+
+void
+openvasd_vt_single_free (openvasd_vt_single_t *);
+void
+
+openvasd_vt_single_add_value (openvasd_vt_single_t *, const gchar *,
+                              const gchar *);
+
+/* Scan config builder */
+char *
+openvasd_build_scan_config_json (openvasd_target_t *, GHashTable *, GSList *);
+
+/* Curl multiperform wrapper */
+typedef struct openvasd_curlm openvasd_curlm_t;
+
+/** @brief Define a string struct for storing the response.
+ */
+typedef struct openvasd_string
+{
+  gchar *ptr;
+  size_t len;
+} openvasd_stringstream;
+
+void
+init_openvasd_stringstream (openvasd_stringstream *s);
+
+void
+reset_openvasd_stringstream (openvasd_stringstream *s);
+
+void
+free_openvasd_stringstream (openvasd_stringstream *s);
```

Comment:
I think it's a bit risky using `free` here. It makes it look like you can allocate an openvasd_stringstream and then free it with this function.

Also the prefix naming breaks the pattern in the file.  Maybe `openvasd_string_cleanup`.

(I do have the feeling this should just be handled inside an openvasd_response, but you understand better how this will be used.)

Reply:
> I do have the feeling this should just be handled inside an openvasd_response, but you understand better how this will be used.

You can take a look into https://github.com/greenbone/gvmd/pull/2262 . In manage.c is use the easy handler, but in manage_sql_nvts.c it is use multiperform handler which receives the chunks of Feed, and they are used to feed the json pull parser. Please, take in account that the gvmd PR is still a draft. However, feel free to start a review ;) (thanks in advance).

Reply:
OK I looked more at 2262.
 - The stringstream and curl handle are always created together so they can be one structure.
 - "curl" is really an internal detail, so the structure would be better as `openvasd_vts_stream_t`.
 - In English a stream of VTs is called a VT stream, so the s can be left out: `openvasd_vt_stream_t`.
 - The structure can be stack allocated, so the `_new` can be skipped entirely.

That leaves the caller looking like this:
```c
  openvasd_vt_stream_t vts;
  resp = openvasd_vt_stream_init (connector, &vts);
  //...
  running = openvasd_vt_stream_get (&vts);
  fwrite (openvasd_vt_stream_str(&vts), 1, openvasd_vt_stream_len(&vts), stream);
  openvasd_vt_stream_reset (&vts);
  int break_flag = 0;
  while (running)
    {
      while (running > 0 && openvasd_vt_stream_len(&vts) < GVM_JSON_PULL_READ_BUFFER_SIZE * 8)
        running = openvasd_vt_stream_get (&vts);
  //...
  openvasd_vt_stream_cleanup (&vts);
```

Does that help to show how I think this can be streamlined?

Note that I think the above is still too heavyweight and the `openvasd_vt_stream` should be kept internally in the connector. So you would call `resp = openvasd_vt_stream_init (connector)` and it would init a field for the VT stream on the connect struct. Then you would call `openvasd_vt_stream_get(connector)` etc. And it would all be freed when you free the connector.

Reply:
Now, the curl handler and the string stream are hidden in the connector struct. So, everything is cleaned up when freeing the the connector. Also, the string is only accessible via functions.
Thanks for the hints and the review!